### PR TITLE
rpc/wallet: Add details and duplicate section for simulaterawtransaction

### PR DESCRIPTION
**Pull request by anibilthare**
The currently provided solution implements a RPC that iterates over the provided list of raw transactions and provides users with the balance change that the transactions will cause in the wallet but the current implementation does not deal with the conflicts that might arise while executing the transactions. This implementation is a follow up for the same, this implementation provides following enhancements to the RPC
            - For every transaction, say tx1, in the input list, a list of unconfirmed transactions in the wallet which conflict with tx1
            - For every transaction, say tx2, in the input list, a list of transactions in the same list which conflict with tx2
            - In case the user provides a transaction more than once, after the first occurrence, the rest of the occurrences are ignored and the index(0-based) for such transactions is reported to the user
    See also: [PR#22751](https://github.com/bitcoin/bitcoin/pull/22751/)

### CODE REVIEW

File: src/wallet/rpc/wallet.cpp [hunk 1]
```diff
@@ -621,11 +642,8 @@ RPCHelpMan simulaterawtransaction()
     std::shared_ptr<const CWallet> const wallet = GetWalletForJSONRPCRequest(request);
     if (!wallet) return NullUniValue;
     const CWallet* pwallet = wallet.get();
-
```
Comment (achow101) [inline src/wallet/rpc/wallet.cpp]: In bf9deb407a8d73ac3b645b34c9bd6d28729a809c "rpc/wallet: Add details and duplicate section for simulaterawtransaction"

nit: Unrelated whitespace change

File: src/wallet/rpc/wallet.cpp [hunk 2]
```diff
@@ -621,11 +642,8 @@ RPCHelpMan simulaterawtransaction()
...
     RPCTypeCheck(request.params, {UniValue::VARR, UniValue::VOBJ}, true);
-
```
Comment (achow101) [inline src/wallet/rpc/wallet.cpp]: In bf9deb407a8d73ac3b645b34c9bd6d28729a809c "rpc/wallet: Add details and duplicate section for simulaterawtransaction"

nit: Unrelated whitespace change

File: src/wallet/rpc/wallet.cpp [hunk 3]
```diff
@@ -621,11 +642,8 @@ RPCHelpMan simulaterawtransaction()
...
     LOCK(pwallet->cs_wallet);
-
```
Comment (achow101) [inline src/wallet/rpc/wallet.cpp]: In bf9deb407a8d73ac3b645b34c9bd6d28729a809c "rpc/wallet: Add details and duplicate section for simulaterawtransaction"

nit: Unrelated whitespace change

File: src/wallet/rpc/wallet.cpp [hunk 4]
```diff
@@ -637,68 +655,136 @@ RPCHelpMan simulaterawtransaction()
 
         include_watchonly = options["include_watchonly"];
     }
-
```
Comment (achow101) [inline src/wallet/rpc/wallet.cpp]: In bf9deb407a8d73ac3b645b34c9bd6d28729a809c "rpc/wallet: Add details and duplicate section for simulaterawtransaction"

nit: Unrelated whitespace change

File: src/wallet/rpc/wallet.cpp [hunk 5]
```diff
@@ -637,68 +655,136 @@ RPCHelpMan simulaterawtransaction()
...
-            if (new_utxos.count(op)) {
-                changes -= new_utxos.at(op);
-                new_utxos.erase(op);
-            } else {
-                if (!coins.count(op)) {
-                    throw JSONRPCError(RPC_INVALID_PARAMETER, "One or more transaction inputs are missing");
-                }
-                if (coins.at(op).IsSpent()) {
-                    throw JSONRPCError(RPC_INVALID_PARAMETER, "One or more transaction inputs have been spent already");
+            // if user tries to double spend in the provided transactions themselves then without this condition we will get an inappropirate error
```
Comment (achow101) [inline src/wallet/rpc/wallet.cpp:703]: In bf9deb407a8d73ac3b645b34c9bd6d28729a809c "rpc/wallet: Add details and duplicate section for simulaterawtransaction"

This comment is a little confusing to understand.

``` suggestion
            // Only apply changes the first time a transaction is seen. Otherwise duplicate transactions will give incorrect errors.
```

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -556,19 +555,26 @@ std::set<uint256> CWallet::GetConflicts(const uint256& txid) const
...
+    for (const CTxIn& txin : tx.vin) {
+        if (mapTxSpends.count(txin.prevout) <= (tx_is_in_wallet ? 1 : 0))
+            continue;  // No conflict if number of tx spending given output <= tx_is_in_wallet
         range = mapTxSpends.equal_range(txin.prevout);
         for (TxSpends::const_iterator _it = range.first; _it != range.second; ++_it)
             result.insert(_it->second);
     }
     return result;
 }
-
```
Comment (achow101) [inline src/wallet/wallet.cpp:571]: In bf9deb407a8d73ac3b645b34c9bd6d28729a809c "rpc/wallet: Add details and duplicate section for simulaterawtransaction"

nit: Unrelated whitespace change

File: src/wallet/wallet.cpp [hunk 7]
```diff
@@ -545,7 +545,6 @@ void CWallet::SetMinVersion(enum WalletFeature nVersion, WalletBatch* batch_in)
             delete batch;
     }
 }
-
```
Comment (achow101) [inline src/wallet/wallet.cpp:548]: In bf9deb407a8d73ac3b645b34c9bd6d28729a809c "rpc/wallet: Add details and duplicate section for simulaterawtransaction"

nit: Unrelated whitespace change

File: src/wallet/wallet.cpp [hunk 8]
```diff
@@ -556,19 +555,26 @@ std::set<uint256> CWallet::GetConflicts(const uint256& txid) const
         return result;
     const CWalletTx& wtx = it->second;
 
+    result = GetConflicts(*(wtx.tx), true);
+
+    return result;
+}
+// tx_is_in_wallet being true means the transaction belongs to wallet and tx_is_in_wallet being false means it is an external transaction
+std::set<uint256> CWallet::GetConflicts(const CTransaction& tx, bool tx_is_in_wallet) const
```
Comment (achow101) [inline src/wallet/wallet.cpp:563]: In bf9deb407a8d73ac3b645b34c9bd6d28729a809c "rpc/wallet: Add details and duplicate section for simulaterawtransaction"

This refactor of `GetConflicts` could be in a separate commit that precedes this commit.

File: src/wallet/wallet.cpp [hunk 9]
```diff
@@ -556,19 +555,26 @@ std::set<uint256> CWallet::GetConflicts(const uint256& txid) const
         return result;
     const CWalletTx& wtx = it->second;
 
+    result = GetConflicts(*(wtx.tx), true);
+
+    return result;
```
Comment (achow101) [inline src/wallet/wallet.cpp:560]: In bf9deb407a8d73ac3b645b34c9bd6d28729a809c "rpc/wallet: Add details and duplicate section for simulaterawtransaction"

`result` is not needed here.

```suggestion
    return Getconflicts(*wtx.tx, true);
````

File: test/functional/wallet_simulaterawtx.py [hunk 10]
```diff
@@ -46,41 +46,87 @@ def run_test(self):
 
         # Add address1 as watch-only to w2
         w2.importpubkey(pubkey=w1.getaddressinfo(address1)["pubkey"])
-
```
Comment (achow101) [inline test/functional/wallet_simulaterawtx.py]: In bf9deb407a8d73ac3b645b34c9bd6d28729a809c "rpc/wallet: Add details and duplicate section for simulaterawtransaction"

nit: Unrelated whitespace change

File: test/functional/wallet_simulaterawtx.py [hunk 11]
```diff
@@ -46,41 +46,87 @@ def run_test(self):
...
         tx1 = node.createrawtransaction([], [{address1: 5.0}])
         tx2 = node.createrawtransaction([], [{address2: 10.0}])
 
         # w0 should be unaffected, w2 should see +5 for tx1
         assert_equal(w0.simulaterawtransaction([tx1])["balance_change"], 0.0)
         assert_equal(w2.simulaterawtransaction([tx1])["balance_change"], 5.0)
-
```
Comment (achow101) [inline test/functional/wallet_simulaterawtx.py]: In bf9deb407a8d73ac3b645b34c9bd6d28729a809c "rpc/wallet: Add details and duplicate section for simulaterawtransaction"

nit: Unrelated whitespace change

File: test/functional/wallet_simulaterawtx.py [hunk 12]
```diff
@@ -46,41 +46,87 @@ def run_test(self):
...
+        assert_equal(w1.simulaterawtransaction([tx1, tx2]), rpc_output)
 
         # w0 funds transaction; it should now see a decrease in (tx fee and payment), and w1 should see the same as above
         funding = w0.fundrawtransaction(tx1)
         tx1 = funding["hex"]
         tx1changepos = funding["changepos"]
         bitcoin_fee = Decimal(funding["fee"])
-
-        # w0 sees fee + 5 btc decrease, w2 sees + 5 btc
+        # w0 sees (fee + 5) btc decrease, w2 sees + 5 btc
```
Comment (achow101) [inline test/functional/wallet_simulaterawtx.py:108]: In bf9deb407a8d73ac3b645b34c9bd6d28729a809c "rpc/wallet: Add details and duplicate section for simulaterawtransaction"

Unrelated change

File: test/functional/wallet_simulaterawtx.py [hunk 13]
```diff
@@ -125,5 +167,183 @@ def run_test(self):
         assert_raises_rpc_error(-8, "One or more transaction inputs have been spent already", w1.simulaterawtransaction, [tx1, tx2])
         assert_raises_rpc_error(-8, "One or more transaction inputs have been spent already", w2.simulaterawtransaction, [tx1, tx2])
 
+
+
```
Comment (achow101) [inline test/functional/wallet_simulaterawtx.py:171]: In bf9deb407a8d73ac3b645b34c9bd6d28729a809c "rpc/wallet: Add details and duplicate section for simulaterawtransaction"

nit: Extra blank lines can be removed.

File: test/functional/wallet_simulaterawtx.py [hunk 14]
```diff
@@ -125,5 +167,183 @@ def run_test(self):
...
+                "txid" : tx2ob["txid"],
+                "wallet_conflicts" : [tx2ob["txid"]],
+                "simulated_tx_conflicts" : []
+            },
+        ]
+        sim_result = w0.simulaterawtransaction([tx2, tx2, tx2])
+        assert_equal(sim_result["details"], details)
+        assert_equal(sim_result["duplicates"], duplicates)
+
+
```
Comment (achow101) [inline test/functional/wallet_simulaterawtx.py:199]: In bf9deb407a8d73ac3b645b34c9bd6d28729a809c "rpc/wallet: Add details and duplicate section for simulaterawtransaction"

nit: Extra blank lines can be removed.

File: test/functional/wallet_simulaterawtx.py [hunk 15]
```diff
@@ -125,5 +167,183 @@ def run_test(self):
...
+            {
+                "txid" : tx4ob["txid"],
+                "wallet_conflicts" : [],
+                "simulated_tx_conflicts" : []
+            },
+        ]
+        sim_result = w0.simulaterawtransaction([tx2, tx3, tx4])
+        assert_equal(sim_result["details"], details)
+        assert_equal(sim_result["duplicates"], [])
+
```
Comment (achow101) [inline test/functional/wallet_simulaterawtx.py:224]: In bf9deb407a8d73ac3b645b34c9bd6d28729a809c "rpc/wallet: Add details and duplicate section for simulaterawtransaction"

nit: Extra blank lines can be removed.

File: src/wallet/rpc/wallet.cpp [hunk 16]
```diff
@@ -610,6 +610,27 @@ RPCHelpMan simulaterawtransaction()
...
             {
                 {RPCResult::Type::STR_AMOUNT, "balance_change", "The wallet balance change (negative means decrease)."},
+                {RPCResult::Type::ARR, "details", "",
+                {
+                    {RPCResult::Type::OBJ, "", "",
+                        {
+                            {RPCResult::Type::STR_HEX, "txid", "the transaction hex of transaction provided by the user"},
+                            {RPCResult::Type::ARR, "wallet_conflicts", "list of txids of transactions in the wallet that conflict with this transaction",
+                            {
+                                {RPCResult::Type::STR_HEX, "txid_in_wallet", "the txid of conflicting transactions in wallet"},
```
Comment (achow101) [inline src/wallet/rpc/wallet.cpp:620]: In bf9deb407a8d73ac3b645b34c9bd6d28729a809c "rpc/wallet: Add details and duplicate section for simulaterawtransaction"

`txid` is sufficient

```suggestion
                                {RPCResult::Type::STR_HEX, "txid", "the txid of conflicting transactions in wallet"},
```

**Review by achow101 — COMMENTED** [review]
Note: @anibilthare is my Summer of Bitcion mentee.

**Comment by achow101** [comment]
Can be rebased now that #22751 is merged.

**Review by josibake — COMMENTED** [review]
Concept ACK

Nice work with informing the user of conflicts! Seems to definitely improve the usability of the RPC. I tried rebasing on master and the rebase wasn't trivial, so I'll wait until you're able to rebase to do a more thorough code review. I left a few comments on the design of the `details` object and a few nits on the code. In general, I'd avoid making stylistic changes (adding/removing whitespace) and avoid adding/editing comments on the existing code.

Lastly, it might help to break this up into a few smaller commits to make it easier to review. I'd suggest putting the refactors (renaming variables, refactoring `GetConflicts`) into their own refactor commit and a separate commit for the new functionality.

File: src/wallet/rpc/wallet.cpp [hunk 17]
```diff
@@ -610,6 +610,27 @@ RPCHelpMan simulaterawtransaction()
...
+                            {RPCResult::Type::ARR, "simulated_tx_conflicts", "indices of transactions provided by the user that conflict with this transaction",
+                            {
+                                {RPCResult::Type::NUM, "index", "index of conflicting transactions in list provided by user"},
+                            }},
+                        }
+                    },
+                }},
+                {RPCResult::Type::ARR, "duplicates", "indices of transactions provided by the user more than once and hence ignored",
+                {
+                    {RPCResult::Type::NUM, "index", "index of transactions provided by the user more than once"},
```
[the remark this answers is not in the document]
Comment (anibilthare) [reply src/wallet/rpc/wallet.cpp:631]: I did consider what you suggested but let's say we do that, then the user will remove those duplicate transactions from the list and re run the RPC? To me this seemed redundant and that's why I took this approach.

File: src/wallet/rpc/wallet.cpp [hunk 18]
```diff
@@ -610,6 +610,27 @@ RPCHelpMan simulaterawtransaction()
...
+                        {
+                            {RPCResult::Type::STR_HEX, "txid", "the transaction hex of transaction provided by the user"},
+                            {RPCResult::Type::ARR, "wallet_conflicts", "list of txids of transactions in the wallet that conflict with this transaction",
+                            {
+                                {RPCResult::Type::STR_HEX, "txid", "the txid of conflicting transactions in wallet"},
+                            }},
+                            {RPCResult::Type::ARR, "simulated_tx_conflicts", "indices of transactions provided by the user that conflict with this transaction",
+                            {
+                                {RPCResult::Type::NUM, "index", "index of conflicting transactions in list provided by user"},
+                            }},
```
[the remark this answers is not in the document]
Comment (anibilthare) [reply src/wallet/rpc/wallet.cpp:625]: Returning txids increases the "manual" work on user's side, if that makes sense? Consider a case when user provides 50 transactions as input. As an end user I think I'd prefer the software telling me "This transaction of yours conflicts with the transaction 11, 17 and 21 that you provided" rather than "This transaction of yours conflict with the transactions with txids <some-long-string1> <some-long-string2> <some-long-string3>"

**Comment by fanquake** [comment]
@anibilthare are you still working on this? Can you comment where you've addressed the review feedback. Note that at least one test is also failing.
